src/socket/haven_socket.rs

```diff
@@ -81,79 +110,95 @@ impl HavenSocket {
-                );
-                match self
-                    .ctx
-                    .dht_get(endpoint.fingerprint)
-                    .await
-                    .map_err(|_| SocketSendError::DhtError)?
-                {
-                    Some(bob_locator) => {
-                        log::debug!("found rob in the DHT");
-                        let rob = bob_locator.rendezvous_point;
-                        // TODO: encrypt body
-                        // use our N2rSocket to send (msg, endpoint) to Rob
-                        self.n2r_socket
-                            .send_to(fwd_body.into(), Endpoint::new(rob, HAVEN_FORWARD_DOCK))
-                            .await?;
-                        Ok(())
-                    }
-                    None => {
-                        log::debug!("couldn't find {} in the DHT", endpoint.fingerprint);
-                        Err(SocketSendError::DhtError)
-                    }
-                }
-            }
-        }
+        let enc = self.encrypters.get_with(endpoint, || {
+            Encrypter::client_new(
+                self.identity_sk,
+                endpoint,
+                self.rendezvous_point,
+                self.n2r_socket.clone(),
+                self.send_incoming_decrypted.clone(),
+                self.ctx.clone(),
+            )
+        });
+        if let Err(e) = enc.send_outgoing(body).await {
+            self.encrypters.remove(&endpoint);
+            log::warn!("Encrypter for {endpoint} FAILED with ERR: {e}! Removed from cache");
+        };
+        Ok(())
     }
 
     pub async fn recv_from(&self) -> Result<(Bytes, Endpoint), SocketRecvError> {
-        let (n2r_msg, _) = self.n2r_socket.recv_from().await?;
-        // TODO: decrypt
-        let inner =
-            stdcode::deserialize(&n2r_msg).map_err(|_| SocketRecvError::HavenMsgBadFormat)?;
-        Ok(inner)
+        self.recv_incoming_decrypted
+            .recv_blocking()
+            .map_err(|_| SocketRecvError::HavenRecvError)
     }
 
     pub fn local_endpoint(&self) -> Endpoint {
         self.n2r_socket.local_endpoint()
     }
 }
+
+async fn recv_task(
+    n2r_skt: N2rSocket,
+    encrypters: Cache<Endpoint, Encrypter>,
+    isk: IdentitySecret,
+    rob: Option<Fingerprint>,
+    send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+    ctx: DaemonContext,
+) -> anyhow::Result<()> {
+    loop {
+        let (n2r_msg, _rendezvous_ep) = n2r_skt.recv_from().await?;
+        let (body, remote_ep): (Bytes, Endpoint) = stdcode::deserialize(&n2r_msg)?;
+        let haven_msg: HavenMsg = stdcode::deserialize(&body)?;
+
+        let encrypter = encrypters.get(&remote_ep);
+        match haven_msg.clone() {
+            HavenMsg::ServerHs(_) => match encrypter {
+                Some(enc) => enc.send_incoming(haven_msg).await?,
+                None => anyhow::bail!("stray msg; dropping"),
+            },
+            HavenMsg::ClientHs(hs) => encrypters.insert(
```

Comment:
Hmm, so here I think there's a design problem --- whether or not the `ClientHs` is valid, we always create a new `Encrypter`, which may overwrite an existing one. But this allows random people to send us messages, claim they come from some particular source, and overwrite the `Encrypter` with something useless.

Of course, they are unable to actually finish establishing a connection, or read any secrets, but they will bump the existing `Encrypter` out, which can cause denial of service.

A possible fix is to have the `Encrypter` server-side constructor be fallible, and verify the handshake.